### PR TITLE
CLDR-14441 Fix DataStoreAdaptorException

### PR DESCRIPTION
**Pull request by srl295**
CLDR-14441

- use a ManagedThreadPool (and friends) instead of using our own threads
- cleanup some SQL use

Remaining warnings:

```
WLTC0032W: Uno o más recursos de transacciones locales se han retrotraído durante la limpieza de un LocalTransactionContainment.
WLTC0033W: El recurso dataSource[DefaultDataSource]/connectionManager se ha retrotraído en la limpieza de LocalTransactionContainment.
```

- made VV single threaded, filed CLDR-14543 to reevaluate

### CODE REVIEW

**Comment by srl295** [comment]
for rationale, see : https://unicode-org.atlassian.net/browse/CLDR-14441

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java [hunk 1]
```diff
@@ -139,23 +131,27 @@ public Appendable statsShort(Appendable output) throws IOException {
         return output.append("" + db_number_open).append("/" + db_max_open);
     }
 
+    public static boolean DO_CLOSE = true;
+
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java:135]: final?

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java [hunk 2]
```diff
@@ -139,23 +131,27 @@ public Appendable statsShort(Appendable output) throws IOException {
...
             }
-            try {
-                if (db_Derby /*&&
-                    datasource instanceof EmbeddedDataSource*/ &&
-                    !conn.isClosed() &&
-                    !conn.getAutoCommit()) {
-                    // commit on close if we are using Derby directly
-                    conn.commit();
+            if(DO_CLOSE) {
+                try {
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java:142]: formatting

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java [hunk 3]
```diff
@@ -601,12 +620,23 @@ public boolean isBogus() {
...
+     */
+    static InitialContext getInitialContext() {
+        InitialContext context;
+        try {
+            context = new InitialContext();
+            return context;
+        } catch (NamingException e) {
+            SurveyMain.busted("Could not get InitialContext", e);
+            // TODO Auto-generated catch block
+            e.printStackTrace();
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java:635]: TODO Auto-generated catch block

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java [hunk 4]
```diff
@@ -618,59 +648,57 @@ private DBUtils() {
...
-                    datasource = myOtherDB;
-                } else {
-                    System.err.println("DefaultDataSource was not OK");
-                    Context initialContext = new InitialContext();
-                    Context eCtx = (Context) initialContext.lookup("java:comp/env");
-                    datasource = (DataSource) eCtx.lookup(JDBC_SURVEYTOOL);
-                }
+            if(myOtherDB != null) {
+                System.err.println("DefaultDataSource was ok");
+                datasource = myOtherDB;
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java:656]: could we avoid using System.err for non-errors?

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java [hunk 5]
```diff
@@ -618,59 +648,57 @@ private DBUtils() {
...
+                    + JDBC_SURVEYTOOL + "'" + ".getConnection() returns : " + t.toString() + "\n" + unchainSqlException(t));
+            } finally {
+                if (c != null)
+                    try {
+                    c.close();
+                    } catch (Throwable tt) {
+                    System.err.println("Couldn't close datasource's conn: " + tt.toString());
+                    tt.printStackTrace();
+                    }
             }
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java:689]: formatting

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyThreadManager.java [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+            if(CLDRConfig.getInstance().getEnvironment() == Environment.UNITTEST) {
+                gScheduledExecutor = Executors.newScheduledThreadPool(2);
+            } else {
+                InitialContext context = DBUtils.getInitialContext();
+                ManagedScheduledExecutorService service = null;
+                try {
+                    service = (ManagedScheduledExecutorService) context.lookup(DEFAULT_MANAGED_SCHEDULED_EXECUTOR);
+                } catch (NamingException e) {
+                    // TODO Auto-generated catch block
+                    SurveyMain.busted("Could not look up " + DEFAULT_MANAGED_SCHEDULED_EXECUTOR, e);
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyThreadManager.java:97]: TODO Auto-generated catch block

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/VettingViewerQueue.java [hunk 7]
```diff
@@ -351,6 +281,87 @@ public void done() {
...
+
+            vv = new VettingViewer<>(sm.getSupplementalDataInfo(), sm.getSTFactory(),
+                getUsersChoice(sm), "Winning " + SurveyMain.getNewVersion());
+            progress.update("Got VettingViewer");
+            statusCode = Status.PROCESSING;
+            start = System.currentTimeMillis();
+            last = start;
+            n = 0;
+            vv.setProgressCallback(new VettingViewer.ProgressCallback() {
+                public String setRemStr(long now) {
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/VettingViewerQueue.java:296]: processCriticalWork is too long and deeply indented; how about moving setProgressCallback to a subroutine

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/VettingViewerQueue.java [hunk 8]
```diff
@@ -351,6 +281,87 @@ public void done() {
...
+                vv.generateHtmlErrorTables(aBuffer, choiceSet, locale.getBaseName(), usersOrg, usersLevel, false);
+            } else {
+                if (DEBUG)
+                    System.err.println("Starting summary gen..");
+                /*
+                 * TODO: remove call to getLocalesWithVotes here, unless it has necessary side-effects.
+                 * Its return value was formerly an unused argument to generateSummaryHtmlErrorTables
+                 */
+                getLocalesWithVotes(st_org);
+                vv.generateSummaryHtmlErrorTables(aBuffer, choiceSet, usersOrg);
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/VettingViewerQueue.java:357]: does getLocalesWithVotes have necessary side-effects?

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyThreadManager.java [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    private static final String JAVA_COMP_DEFAULT_MANAGED_THREAD_FACTORY = "java:comp/DefaultManagedThreadFactory";
+    private static final String DEFAULT_MANAGED_EXECUTOR = "java:comp/DefaultManagedExecutorService";
+    private static final String DEFAULT_MANAGED_SCHEDULED_EXECUTOR = "java:comp/DefaultManagedScheduledExecutorService";
+    static ThreadFactory gFactory = null;
+
+    static ExecutorService gExecutor = null;
+    static ScheduledExecutorService gScheduledExecutor = null;
+
+    public static ThreadFactory getThreadFactory() {
+        if(gFactory == null) {
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyThreadManager.java:41]: does getThreadFactory need to be public? it only seems to be called locally

also the callers are both lower in the file, so maybe this should be private and moved below the callers

Comment (btangmu) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyThreadManager.java:41]: gExecutor could also be static?

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java (hunk 1 above)
Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java:135]: should have been deleted.  This was from incorrectly backing out an experiment, I'll refactor this file.

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java (hunk 4 above)
Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java:656]: Actually, we should be using a real logger. I thought about this the other day, and I see there's already an issue filed for it: https://unicode-org.atlassian.net/browse/CLDR-8581

Anyway, I will correct some of these.

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java (hunk 5 above)
Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/DBUtils.java:689]: fixed

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/VettingViewerQueue.java (hunk 7 above)
Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/VettingViewerQueue.java:296]: actually the anonymous class is what needs to have a name.

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/VettingViewerQueue.java (hunk 8 above)
Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/VettingViewerQueue.java:357]: I can't see any. THis should be a separate ticket, though.

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyThreadManager.java (hunk 9 above)
Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyThreadManager.java:41]: So, I'd like the ThreadFactory to be the preferred API (being standard) , and get rid of the homegrown addTask API. I'll do that.

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyThreadManager.java (hunk 6 above)
Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyThreadManager.java:97]: fixed

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/VettingViewerQueue.java (hunk 8 above)
Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/VettingViewerQueue.java:357]: Filed https://unicode-org.atlassian.net/browse/CLDR-14537

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyThreadManager.java (hunk 9 above)
Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyThreadManager.java:41]: done

**Review by btangmu — APPROVED** [review]

**Pull request merged**